src/main/java/hudson/plugins/deploy/DeployPublisher.java

```diff
@@ -48,18 +55,49 @@ public DeployPublisher(List<ContainerAdapter> adapters, String war, String conte
     }
 
     @Override
+    @Deprecated
     public boolean perform(AbstractBuild<?,?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
-        if (build.getResult().equals(Result.SUCCESS) || onFailure) {
-            for (FilePath warFile : build.getWorkspace().list(this.war)) {
-                for (ContainerAdapter adapter : adapters)
-                    if (!adapter.redeploy(warFile, contextPath, build, launcher, listener))
-                        build.setResult(Result.FAILURE);
-            }
+        FilePath workspace = build.getWorkspace();
+        if (workspace  == null) {
+            listener.getLogger().println("[DeployPublisher][ERROR] Workspace not found");
+            throw new AbortException("Workspace not found");
         }
-
+        perform(build, workspace, launcher, listener);
         return true;
     }
 
+    @Override
+    public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnull Launcher launcher, @Nonnull TaskListener listener) throws InterruptedException, IOException {
+        Result result = run.getResult();
+        if (onFailure || Result.SUCCESS.equals(result)) {
```

Comment:
🐛 If `onFailure` is set, you will go into this branch even if the build is running.
I would also expect `result == null` to go here since you expect the deployment to happen for the running Pipeline